crypto/dilithium/internal.h

```diff
@@ -0,0 +1,66 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#ifndef AWSLC_HEADER_SIG_INTERNAL_H
+#define AWSLC_HEADER_SIG_INTERNAL_H
+
+#include <openssl/base.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+// NISTDSA_METHOD structure and helper functions.
+typedef struct {
```

Comment:
Yes, I'm currently writing up a document that describes how pre-hash modes will integrate into this API.

Essentially,
```
// NISTDSA_METHOD structure and helper functions.
typedef struct {
  int (*keygen)(uint8_t *public_key,
                uint8_t *secret_key);

  int (*sign)(uint8_t *sig, 
              size_t *sig_len,
              const uint8_t *message,
              size_t message_len,
              const uint8_t *ctx,
              size_t ctx_len,
              const uint8_t *secret_key);
              
  int (*hashsign)(uint8_t *sig, 
                  size_t *sig_len,
                  const uint8_t *message,
                  size_t message_len,
                  const uint8_t *ctx,
                  size_t ctx_len,
                  const uint8_t *secret_key
                  int hash_OID);

  int (*verify)(const uint8_t *message,
                size_t message_len,
                const uint8_t *sig,
                size_t sig_len,
                const uint8_t *ctx,
                size_t ctx_len,
                const uint8_t *public_key);

} NISTDSA_METHOD;
```

```
DEFINE_LOCAL_DATA(NISTDSA_METHOD, pre_hash_sig_ml_dsa_65_method) {
  out->keygen = ml_dsa_65_keypair;
  out->hashsign = hash_ml_dsa_65_sign;
  out->verify = hash_ml_dsa_65_verify;
}
```

```
int hash_ml_dsa_65_sign(uint8_t *sig               /* OUT */,
                        size_t *sig_len            /* OUT */,
                        const uint8_t *message     /* IN */,
                        size_t message_len         /* IN */,
                        const uint8_t *ctx         /* IN */,
                        size_t ctx_len             /* IN */,
                        const uint8_t *secret_key  /* IN */,
                        int hash_OID) {
  ml_dsa_params params;
  ml_dsa_65_params_init(&params);
  return pre_hash_crypto_sign_signature(&params, sig, sig_len, message, message_len,
                                         ctx, ctx_len, secret_key, hash_OID);
}
```
Where `pre_hash_crypto_sign_signature` is [FIPS 204: Algorithm 4](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.204.pdf) `HashML-DSA.Sign(sk, M, ctx, PH)`. This then calls `ML-DSA.Sign_internal`.

Reply:
We can flesh this out in a full design when the pre-hash modes are added. My rational was that since NIST has provided separate OIDs for pre-hash modes:

```
HashML-DSA-44 : 2.16.840.1.101.3.4.3.32
HashML-DSA-65 : 2.16.840.1.101.3.4.3.33
HashML-DSA-87 : 2.16.840.1.101.3.4.3.34
```

then we create a new method, say `pre_hash_sig_ml_dsa_65` that can signs **only** in pre-hash modes:

```
DEFINE_LOCAL_DATA(NISTDSA, pre_hash_sig_ml_dsa_65) {
  out->nid = NID_HashMLDSA65;
  out->oid = kOID_HashMLDSA65;
  out->oid_len = sizeof(kOID_HashMLDSA65);
  out->comment = "HashMLDSA65 ";
  out->public_key_len = MLDSA65_PUBLIC_KEY_BYTES;
  out->secret_key_len = MLDSA65_PRIVATE_KEY_BYTES;
  out->signature_len = MLDSA65_SIGNATURE_BYTES;
  out->keygen_seed_len = MLDSA65_KEYGEN_SEED_BYTES;
  out->sign_seed_len = MLDSA65_SIGNATURE_SEED_BYTES;
  out->method = pre_hash_sig_ml_dsa_65_method();
}
```

this way we connect the correct OID (`NID_HashMLDSA65`) to the method.

If we were to add pre-hash and pure signatures under the same method, we'd have to add the abstraction between the two modes in the EVP layer, which seems more complicated to me. If NIST sees this as a different OID, I see it as a new method. But again, we can loop back on this when pre-hash modes are added in another PR.

Reply:
ok, just wanted to check it's on your radar, we can discuss this when you get to it.